feat: debug options (#1160)
callstack/react-native-testing-library: website/docs/API.md

### `debug`

```ts

interface DebugOptions { 
  message?: string;
  filterProps?: FilterNodeByTypeFunction;
}

> debug(options?: string | DebugOptions): void

Reviewer: ```suggestion
debug(options?: DebugOptions | string): void
```

```

Pretty prints deeply rendered component passed to `render`. 

You can provide a message that will be printed on top either by calling debug with a string or using the message property in debugOptions

```jsx
render(<Component />);

> screen.debug('optional message');

Reviewer: Let's suggest them only the modern way.
```suggestion
```

> screen.debug({message: 'optional message'});

Reviewer: formatting:
```suggestion
screen.debug({ message: 'optional message' });
```

### `debug`

```ts

interface DebugOptions { 
  message?: string;
  mapProps?: MapPropsFunction;
}

debug(options?: DebugOptions | string): void
```

Pretty prints deeply rendered component passed to `render`. 

> 

Reviewer: Pls add lower-level headers for options as we do for other APIs

```suggestion
#### `message` option
```

Author: Very good idea I hadn't noticed this was done for other apis !

</View>
```

You can use the filterProps option to filter the props that will be printed : 

```jsx
render(<View style={{backgroundColor: 'red'}}/>);

debug({filterProps : (propName, propValue, node) => propName !== 'style'})
```

> This will log the rendered JSX without style props

Reviewer: ```suggestion
This will log the rendered JSX without the `style` prop.
```

</View>
```

> 

Reviewer: ```suggestion
#### `mapProps` option
```

> You can use the mapProps option to transform the props that will be printed : 

Reviewer: ```suggestion
You can use the `mapProps` option to transform the props that will be printed : 
```

```jsx

> render(<View style={{backgroundColor: 'red'}}/>);
> debug({ mapProps : ({style, ...props}) => ({props}) })

Reviewer: formatting:
```suggestion
render(<View style={{ backgroundColor: 'red' }}/>);
debug({ mapProps : ({style, ...props}) => ({ props }) })
```

```

This will log the rendered JSX without the `style` props. 

> This option can be used to target specfic props when debugging a query (for instance keeping only children prop when debugging a byText query) and you can also transform prop values so that they are more readable (e.g. flatten styles).

Reviewer: Not sure about mentioning `children` prop in this case, as it is treated specially by `mapProps`

```suggestion
This option can be used to target specific props when debugging a query (for instance keeping only `children` prop when debugging a `getByText` query) and you can also transform prop values so that they are more readable (e.g. flatten styles).
```

Author: I changed the order so that it is explained first that children prop cannot be filtered then give the example of usage, since I believe this will be one of the most common, at least until byRole queries become more popular

> The `children` prop cannot be filtered so the following will print all rendered components with all props but children filtered.

Reviewer: ```suggestion
The `children` prop cannot be filtered out so the following will print all rendered components with all props but children filtered.
```

Its value is automatically cleared after each test by calling [`cleanup`](#cleanup). If no `render` call has been made in a given test then it holds a special object that implements `RenderResult` but throws a helpful error on each property and method access.

> This can also be used to build test utils that would normally require to be in render scope, either in a test file or globally for your project. For instance in a test file where we fill a form in multiple tests: 

Reviewer: The concept of building test utils seems like a useful idea, however the `typeEmail` example seems quite artificial and I would not recommend using such util over explicit `fireEvent` + query.

@pierrezimmermannbam do you have idea for something that we could actually recommend to your users here?

Author: This is something is sometimes do when I test forms and have a lot of tests, imo it makes tests more readable and very easy to understand while not making them harder to debug since you still have the same error messages. It's even more interesting when it allows to hide testID usage since the queries by testID do not really help to understand the test. An example of this is waiting for a loader to disapear, which is often done in every test when testing at page level

```tsx
const waitForLoaderToDisapear = async () => {
   await waitForElementToBeRemoved(() => screen.getByTestID('PageLoader'));
}
```

Does this example make more sense to you ?

Default timeout, in ms, for async helper functions (`waitFor`, `waitForElementToBeRemoved`) and `findBy*` queries. Defaults to 1000 ms.

#### `debugOptions` option

> Default debugOptions to be used when calling debug. These default options will be overridden by the ones you specify directly when calling debug

Reviewer: ```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`.
```

Default timeout, in ms, for async helper functions (`waitFor`, `waitForElementToBeRemoved`) and `findBy*` queries. Defaults to 1000 ms.

#### `defaultDebugOptions` option

> Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`. Can be used namely to transform prop values for more readability or filter out props that have little value when debugging tests (e.g. svg's path prop).

Reviewer: I would rather refer user to `debug` `options` docs for details rather than repeating the use cases.

```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`. 
```

Author: agreed, I added a link to debug options doc
